src/coreclr/nativeaot/System.Private.Reflection.Execution/src/Internal/Reflection/Execution/ExecutionEnvironmentImplementation.ManifestResources.cs

```diff
@@ -78,74 +73,37 @@ private static unsafe UnmanagedMemoryStream ReadResourceFromBlob(ResourceInfo re
             return new UnmanagedMemoryStream(pBlob + resourceInfo.Index, resourceInfo.Length);
         }
 
-        private static LowLevelList<ResourceInfo> GetExtractedResources(Assembly assembly)
+        private static IEnumerable<ResourceInfo> ExtractResources(Assembly assembly)
         {
-            LowLevelDictionary<string, LowLevelList<ResourceInfo>> extractedResourceDictionary = ExtractedResourceDictionary;
+            ArgumentNullException.ThrowIfNull(assembly);
             string assemblyName = assembly.GetName().FullName;
-            LowLevelList<ResourceInfo> resourceInfos;
-            if (!extractedResourceDictionary.TryGetValue(assemblyName, out resourceInfos))
-                return new LowLevelList<ResourceInfo>();
-            return resourceInfos;
-        }
 
-        private static LowLevelDictionary<string, LowLevelList<ResourceInfo>> ExtractedResourceDictionary
-        {
-            get
+            foreach (NativeFormatModuleInfo module in ModuleList.EnumerateModules())
             {
-                if (s_extractedResourceDictionary == null)
+                if (!TryGetNativeReaderForBlob(module, ReflectionMapBlob.BlobIdResourceIndex, out NativeReader reader))
                 {
-                    // Lazily create the extracted resource dictionary. If two threads race here, we may construct two dictionaries
-                    // and overwrite one - this is ok since the dictionaries are read-only once constructed and they contain the identical data.
+                    continue;
+                }
+                NativeParser indexParser = new NativeParser(reader, 0);
+                NativeHashtable indexHashTable = new NativeHashtable(indexParser);
 
-                    LowLevelDictionary<string, LowLevelList<ResourceInfo>> dict = new LowLevelDictionary<string, LowLevelList<ResourceInfo>>();
+                var entryEnumerator = indexHashTable.EnumerateAllEntries();
+                NativeParser entryParser;
+                while (!(entryParser = entryEnumerator.GetNext()).IsNull)
```

Comment:
Can we store this in per-assembly metadata so that we do not have to do this lookup using assembly name at all?

Reply:
I tried a bit. `NativeFormatRuntimeAssembly.Scope.MetadataReader._streamReader` has the required information. However, `NativeReader` is a source-cloned type and not exchangeable. Exporting the raw pointer and length works, but looks ugly.